File: app/src/main/java/de/gaffga/android/fragments/SettingsFragment.java
package de.gaffga.android.fragments;

import android.content.DialogInterface;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import androidx.appcompat.app.AlertDialog;
import androidx.preference.CheckBoxPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import android.util.Log;
import android.widget.Toast;
import de.gaffga.android.zazentimer.DbOperations;
import de.gaffga.android.zazentimer.R;
import de.gaffga.android.zazentimer.ZazenTimerActivity;
import de.gaffga.android.zazentimer.ZenTimerDatabase;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.Enumeration;
import java.util.zip.ZipEntry;
import java.util.zip.ZipFile;
import java.util.zip.ZipOutputStream;

public class SettingsFragment extends PreferenceFragmentCompat {
    private static final String TAG = "ZMT_SettingsFragment";
    private static final int REQUEST_BACKUP = 201;
    private static final int REQUEST_RESTORE = 202;

    @Override
    public void onCreatePreferences(Bundle bundle, String rootKey) {
        setPreferencesFromResource(R.xml.preferences, rootKey);

        findPreference(ZazenTimerActivity.PREF_KEY_THEME).setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object obj) {
                requireActivity().runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Intent intent = new Intent(requireActivity(), ZazenTimerActivity.class);
                        intent.putExtra(ZazenTimerActivity.INTENT_DATA_SHOW_PREF_ON_START, true);
                        requireActivity().finish();
                        requireActivity().startActivity(intent);
                    }
                });
                return true;
            }
        });
        final CheckBoxPreference checkBoxPreference = (CheckBoxPreference) findPreference(ZazenTimerActivity.PREF_KEY_OUTPUT_CHANNEL_ALARM);
        final CheckBoxPreference checkBoxPreference2 = (CheckBoxPreference) findPreference(ZazenTimerActivity.PREF_KEY_OUTPUT_CHANNEL_MUSIC);
        checkBoxPreference.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object obj) {
                if (!((Boolean) obj).booleanValue()) {
                    return checkBoxPreference2.isChecked();
                }
                checkBoxPreference2.setChecked(false);
                return true;
            }
        });
        checkBoxPreference2.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object obj) {
                if (!((Boolean) obj).booleanValue()) {
                    return checkBoxPreference.isChecked();
                }
                checkBoxPreference.setChecked(false);
                return true;
            }
        });
        final CheckBoxPreference checkBoxPreference3 = (CheckBoxPreference) findPreference(ZazenTimerActivity.PREF_KEY_MUTE_MODE_VIBRATE_SOUND);
        final CheckBoxPreference checkBoxPreference4 = (CheckBoxPreference) findPreference(ZazenTimerActivity.PREF_KEY_MUTE_MODE_VIBRATE);
        final CheckBoxPreference checkBoxPreference5 = (CheckBoxPreference) findPreference(ZazenTimerActivity.PREF_KEY_MUTE_MODE_NONE);
        CheckBoxPreference checkBoxPreference6 = (CheckBoxPreference) findPreference(ZazenTimerActivity.PREF_KEY_KEEP_SCREEN_ON);
        final Preference brightnessPreference = findPreference(ZazenTimerActivity.PREF_KEY_BRIGHTNESS);
        checkBoxPreference3.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object obj) {
                if (!((Boolean) obj).booleanValue()) {
                    return checkBoxPreference4.isChecked() || checkBoxPreference5.isChecked();
                }
                checkBoxPreference4.setChecked(false);
                checkBoxPreference5.setChecked(false);
                return true;
            }
        });
        checkBoxPreference4.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object obj) {
                if (!((Boolean) obj).booleanValue()) {
                    return checkBoxPreference3.isChecked() || checkBoxPreference5.isChecked();
                }
                checkBoxPreference3.setChecked(false);
                checkBoxPreference5.setChecked(false);
                return true;
            }
        });
        checkBoxPreference5.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object obj) {
                if (!((Boolean) obj).booleanValue()) {
                    return checkBoxPreference3.isChecked() || checkBoxPreference4.isChecked();
                }
                checkBoxPreference3.setChecked(false);
                checkBoxPreference4.setChecked(false);
                return true;
            }
        });
        checkBoxPreference6.setOnPreferenceChangeListener(new Preference.OnPreferenceChangeListener() {
            @Override
            public boolean onPreferenceChange(Preference preference, Object obj) {
                brightnessPreference.setEnabled(((Boolean) obj).booleanValue());
                return true;
            }
        });
        brightnessPreference.setEnabled(checkBoxPreference6.isChecked());

        // Backup via SAF — user picks where to save
        Preference backupPref = findPreference("backup_to_sd");
        backupPref.setEnabled(true);
        backupPref.setSummary(R.string.pref_sum_backup);
        backupPref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                Intent intent = new Intent(Intent.ACTION_CREATE_DOCUMENT);
                intent.addCategory(Intent.CATEGORY_OPENABLE);
                intent.setType("application/zip");
                intent.putExtra(Intent.EXTRA_TITLE, "zazentimer_backup.zip");
                startActivityForResult(intent, REQUEST_BACKUP);
                return true;
            }
        });

        // Restore via SAF — user picks the backup file
        Preference restorePref = findPreference("restore_from_sd");
        restorePref.setEnabled(true);
        restorePref.setSummary(R.string.pref_sum_restore);
        restorePref.setOnPreferenceClickListener(new Preference.OnPreferenceClickListener() {
            @Override
            public boolean onPreferenceClick(Preference preference) {
                new AlertDialog.Builder(requireActivity())
                        .setTitle(R.string.restore_really_title)
                        .setMessage(R.string.restore_really_text)
                        .setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                                Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
                                intent.addCategory(Intent.CATEGORY_OPENABLE);
                                intent.setType("application/zip");
                                startActivityForResult(intent, REQUEST_RESTORE);
                            }
                        })
                        .setNegativeButton(R.string.abbrechen, new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {
                            }
                        })
                        .show();
                return true;
            }
        });
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (resultCode != android.app.Activity.RESULT_OK || data == null) {
            return;
        }
        Uri uri = data.getData();
        if (uri == null) {
            return;
        }
        requireActivity().getContentResolver().takePersistableUriPermission(
                uri, Intent.FLAG_GRANT_READ_URI_PERMISSION | Intent.FLAG_GRANT_WRITE_URI_PERMISSION);

        if (requestCode == REQUEST_BACKUP) {
            doBackup(uri);
        } else if (requestCode == REQUEST_RESTORE) {
            doRestore(uri);
        }
    }

    private void doBackup(Uri uri) {
        final Uri finalUri = uri;
        new AsyncTask<Void, Void, Boolean>() {
            @Override
            protected Boolean doInBackground(Void... voidArr) {
                return Boolean.valueOf(doRealBackup(finalUri));
            }

            @Override
            protected void onPostExecute(Boolean success) {
                if (success.booleanValue()) {
                    Toast.makeText(requireActivity(), R.string.backup_success_text, 0).show();
                } else {
                    Toast.makeText(requireActivity(), R.string.backup_error_text, 0).show();
                }
            }
        }.execute(new Void[0]);
    }

    private boolean doRealBackup(Uri uri) {
        Log.d(TAG, "Backup to URI: " + uri);
        boolean failed = false;
        try {
            OutputStream os = requireActivity().getContentResolver().openOutputStream(uri);
            if (os == null) {
                Log.e(TAG, "Could not open output stream for URI");
                return false;
            }
            ZipOutputStream zos = new ZipOutputStream(os);

            // Backup database
            ZipEntry ze = new ZipEntry("zentimer");
            zos.putNextEntry(ze);
            if (!sendFile(requireActivity().getDatabasePath("zentimer"), zos)) {
                failed = true;
            }
            zos.closeEntry();

            // Backup app files
            File filesDir = requireActivity().getFilesDir();
            File[] listFiles = filesDir.listFiles(f -> !f.getName().equals("InstantRun"));
            if (listFiles != null) {
                for (File file : listFiles) {
                    ZipEntry ze2 = new ZipEntry(file.getName());
                    zos.putNextEntry(ze2);
                    if (!sendFile(file, zos)) {
                        failed = true;
                    }
                    zos.closeEntry();
                }
            }
            zos.close();
        } catch (Exception e) {
            Log.e(TAG, "IO/Error during backup", e);
            failed = true;
        }
        return !failed;
    }

    private void doRestore(Uri uri) {
        final Uri finalUri = uri;
        new AsyncTask<Void, Void, Integer>() {
            @Override
            protected Integer doInBackground(Void... voidArr) {
                return Integer.valueOf(doRealRestore(finalUri));
            }

            @Override
            protected void onPostExecute(Integer num) {
                if (num.intValue() == 0) {
                    Toast.makeText(requireActivity(), R.string.restore_success_text, 0).show();
                } else if (num.intValue() == 1) {
                    Toast.makeText(requireActivity(), R.string.restore_backup_not_found, 0).show();
                } else if (num.intValue() == 2) {
                    Toast.makeText(requireActivity(), R.string.restore_error_text, 0).show();
                }
            }
        }.execute(new Void[0]);
    }

    private int doRealRestore(Uri uri) {
        Log.d(TAG, "Restore from URI: " + uri);
        boolean failed = false;
        try {
            // Copy URI content to a temp file so we can use ZipFile (needs seekable)
            File tempFile = File.createTempFile("restore", ".zip", requireActivity().getCacheDir());
            InputStream is = requireActivity().getContentResolver().openInputStream(uri);
            if (is == null) {
                Log.e(TAG, "Could not open input stream for URI");
                return 1;
            }
            FileOutputStream fos = new FileOutputStream(tempFile);
            byte[] buf = new byte[32768];
            int read;
            while ((read = is.read(buf)) > 0) {
                fos.write(buf, 0, read);
            }
            fos.close();
            is.close();

            ZipFile zipFile = new ZipFile(tempFile);
            Enumeration<? extends ZipEntry> entries = zipFile.entries();
            while (entries.hasMoreElements()) {
                ZipEntry entry = entries.nextElement();
                if (entry.getName().equals(ZenTimerDatabase.DATABASE_NAME)) {
                    DbOperations.close();
                    if (!receiveFile(zipFile.getInputStream(entry), requireActivity().getDatabasePath(ZenTimerDatabase.DATABASE_NAME))) {
                        failed = true;
                    }
                    DbOperations.init(requireActivity());
                } else if (!receiveFile(zipFile.getInputStream(entry), new File(requireActivity().getFilesDir(), entry.getName()))) {
                    failed = true;
                }
            }
            zipFile.close();
            tempFile.delete();
        } catch (Exception e) {
            Log.e(TAG, "Error restoring", e);
            failed = true;
        }
        return failed ? 2 : 0;
    }

    private boolean receiveFile(InputStream inputStream, File file) {
        Log.i(TAG, "receiving File from zip: " + file.getName());
        try {
            FileOutputStream fileOutputStream = new FileOutputStream(file);
            byte[] bArr = new byte[32768];
            int read = inputStream.read(bArr);
            while (read > 0) {
                fileOutputStream.write(bArr, 0, read);
                read = inputStream.read(bArr);
            }
            fileOutputStream.close();
            return true;
        } catch (Exception e) {
            Log.e(TAG, "Error receiving file from zip archive", e);
            return false;
        }
    }

    private boolean sendFile(File file, OutputStream outputStream) {
        Log.i(TAG, "sending File to zip: " + file.getName());
        try {
            FileInputStream fileInputStream = new FileInputStream(file);
            byte[] bArr = new byte[32768];
            for (int read = fileInputStream.read(bArr); read > 0; read = fileInputStream.read(bArr)) {
                outputStream.write(bArr, 0, read);
            }
            fileInputStream.close();
            return true;
        } catch (Exception e) {
            Log.e(TAG, "Error sending file to zip archive", e);
            return false;
        }
    }
}

File: app/src/main/kotlin/at/priv/graf/zazentimer/fragments/SettingsFragment.kt
package at.priv.graf.zazentimer.fragments

import android.app.Activity
import android.content.DialogInterface
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.lifecycleScope
import androidx.preference.CheckBoxPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import at.priv.graf.zazentimer.R
import at.priv.graf.zazentimer.ZazenTimerActivity
import at.priv.graf.zazentimer.backup.BackupManager
import at.priv.graf.zazentimer.database.AppDatabase
import at.priv.graf.zazentimer.database.DbOperations
import com.google.android.material.transition.MaterialFadeThrough
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import javax.inject.Inject

@AndroidEntryPoint
class SettingsFragment : PreferenceFragmentCompat() {
    @Inject
    lateinit var dbOperations: DbOperations

    private val backupManager: BackupManager by lazy {
        BackupManager(
            databaseFileProvider = { requireActivity().getDatabasePath(AppDatabase.DATABASE_NAME) },
            filesDirProvider = { requireActivity().filesDir },
            onCloseDatabase = { dbOperations.close() },
            onReopenDatabase = { dbOperations.reopen() },
        )
    }

    private val backupLauncher: ActivityResultLauncher<Intent> =
        registerForActivityResult(
            ActivityResultContracts.StartActivityForResult(),
        ) { result ->
            if (result.resultCode != Activity.RESULT_OK || result.data == null) return@registerForActivityResult
            val uri = result.data?.data ?: return@registerForActivityResult
            requireActivity().contentResolver.takePersistableUriPermission(
                uri,
                Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION,
            )
            doBackup(uri)
        }

    private val restoreLauncher: ActivityResultLauncher<Intent> =
        registerForActivityResult(
            ActivityResultContracts.StartActivityForResult(),
        ) { result ->
            if (result.resultCode != Activity.RESULT_OK || result.data == null) return@registerForActivityResult
            val uri = result.data?.data ?: return@registerForActivityResult
            requireActivity().contentResolver.takePersistableUriPermission(
                uri,
                Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION,
            )
            doRestore(uri)
        }

    override fun onCreate(bundle: Bundle?) {
        super.onCreate(bundle)
        enterTransition = MaterialFadeThrough()
    }

    override fun onCreatePreferences(
        bundle: Bundle?,
        rootKey: String?,
    ) {
        setPreferencesFromResource(R.xml.preferences, rootKey)
        setupThemePreference()
        setupMuteModePreferences()
        setupBrightnessPreference()
        setupBackupPreferences()
    }

    private fun setupThemePreference() {
        findPreference<Preference>(ZazenTimerActivity.PREF_KEY_THEME)?.setOnPreferenceChangeListener { _, _ ->
            requireActivity().runOnUiThread {
                val intent = Intent(requireActivity(), ZazenTimerActivity::class.java)
                intent.putExtra(ZazenTimerActivity.INTENT_DATA_SHOW_PREF_ON_START, true)
                requireActivity().finish()
                requireActivity().startActivity(intent)
            }
            true
        }
    }

    private fun setupMuteModePreferences() {
        val vibSoundPref =
            findPreference<CheckBoxPreference>(ZazenTimerActivity.PREF_KEY_MUTE_MODE_VIBRATE_SOUND)
        val vibPref =
            findPreference<CheckBoxPreference>(ZazenTimerActivity.PREF_KEY_MUTE_MODE_VIBRATE)
        val nonePref =
            findPreference<CheckBoxPreference>(ZazenTimerActivity.PREF_KEY_MUTE_MODE_NONE)
        if (vibSoundPref == null || vibPref == null || nonePref == null) return
        val checkVibSound = vibSoundPref
        val checkVib = vibPref
        val checkNone = nonePref

        checkVibSound.setOnPreferenceChangeListener { _, obj ->
            if (!(obj as Boolean)) {
                return@setOnPreferenceChangeListener checkVib.isChecked || checkNone.isChecked
            }
            checkVib.isChecked = false
            checkNone.isChecked = false
            true
        }
        checkVib.setOnPreferenceChangeListener { _, obj ->
            if (!(obj as Boolean)) {
                return@setOnPreferenceChangeListener checkVibSound.isChecked || checkNone.isChecked
            }
            checkVibSound.isChecked = false
            checkNone.isChecked = false
            true
        }
        checkNone.setOnPreferenceChangeListener { _, obj ->
            if (!(obj as Boolean)) {
                return@setOnPreferenceChangeListener checkVibSound.isChecked || checkVib.isChecked
            }
            checkVibSound.isChecked = false
            checkVib.isChecked = false
            true
        }
    }

    private fun setupBrightnessPreference() {
        val keepScreenOnPref =
            findPreference<CheckBoxPreference>(ZazenTimerActivity.PREF_KEY_KEEP_SCREEN_ON)
        val brightnessPref =
            findPreference<Preference>(ZazenTimerActivity.PREF_KEY_BRIGHTNESS)
        val checkKeepScreenOn = keepScreenOnPref ?: return
        val checkBrightness = brightnessPref ?: return

        checkKeepScreenOn.setOnPreferenceChangeListener { _, obj ->
            checkBrightness.isEnabled = obj as Boolean
            true
        }
        checkBrightness.isEnabled = checkKeepScreenOn.isChecked
    }

    private fun setupBackupPreferences() {
        val backupPref =
            findPreference<Preference>("backup_to_sd") ?: return
        backupPref.isEnabled = true
        backupPref.setSummary(R.string.pref_sum_backup)
        backupPref.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                val intent = Intent(Intent.ACTION_CREATE_DOCUMENT)
                intent.addCategory(Intent.CATEGORY_OPENABLE)
                intent.type = "application/zip"
                intent.putExtra(Intent.EXTRA_TITLE, "zazentimer_backup.zip")
                backupLauncher.launch(intent)
                true
            }

        val restorePref =
            findPreference<Preference>("restore_from_sd") ?: return
        restorePref.isEnabled = true
        restorePref.setSummary(R.string.pref_sum_restore)
        restorePref.onPreferenceClickListener =
            Preference.OnPreferenceClickListener {
                AlertDialog
                    .Builder(requireActivity())
                    .setTitle(R.string.restore_really_title)
                    .setMessage(R.string.restore_really_text)
                    .setPositiveButton(R.string.ok) { _: DialogInterface, _: Int ->
                        val intent = Intent(Intent.ACTION_OPEN_DOCUMENT)
                        intent.addCategory(Intent.CATEGORY_OPENABLE)
                        intent.type = "application/zip"
                        restoreLauncher.launch(intent)
                    }.setNegativeButton(R.string.abbrechen) { _: DialogInterface, _: Int ->
                    }.show()
                true
            }
    }

    private fun doBackup(uri: Uri) {
        lifecycleScope.launch {
            val os = requireActivity().contentResolver.openOutputStream(uri)
            val success =
                if (os != null) {
                    backupManager.backup(os)
                } else {
                    Log.e(TAG, "Could not open output stream for URI")
                    dbOperations.reopen()
                    false
                }
            if (success) {
                Toast.makeText(requireActivity(), R.string.backup_success_text, 0).show()
            } else {
                Toast.makeText(requireActivity(), R.string.backup_error_text, 0).show()
            }
        }
    }

    private fun doRestore(uri: Uri) {
        lifecycleScope.launch {
            var result = 2
            try {
                val tempFile = File.createTempFile("restore", ".zip", requireActivity().cacheDir)
                val inputStream = requireActivity().contentResolver.openInputStream(uri)
                if (inputStream == null) {
                    Log.e(TAG, "Could not open input stream for URI")
                    result = 1
                } else {
                    val fos = FileOutputStream(tempFile)
                    val buf = ByteArray(BUFFER_SIZE)
                    var read: Int
                    while ((inputStream.read(buf).also { read = it }) > 0) {
                        fos.write(buf, 0, read)
                    }
                    fos.close()
                    inputStream.close()
                    result = backupManager.restore(tempFile)
                    tempFile.delete()
                }
            } catch (e: IOException) {
                Log.e(TAG, "Error restoring", e)
            }
            if (result == 0) {
                Toast.makeText(requireActivity(), R.string.restore_success_text, 0).show()
            } else if (result == 1) {
                Toast.makeText(requireActivity(), R.string.restore_backup_not_found, 0).show()
            } else if (result == 2) {
                Toast.makeText(requireActivity(), R.string.restore_error_text, 0).show()
            }
        }
    }

    companion object {
        private const val TAG = "ZMT_SettingsFragment"
        private const val BUFFER_SIZE = 32768
    }
}
